File: java/hittingtargets.java
import java.util.Scanner;

public class hittingtargets {
	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		int m = scan.nextInt();
		Shape[] s = new Shape[m];
		for (int i = 0; i < m; i++)
			if (scan.next().equals("rectangle"))
				s[i] = new Rectangle(scan.nextInt(), scan.nextInt(), scan.nextInt(), scan.nextInt());
			else
				s[i] = new Circle(scan.nextInt(), scan.nextInt(), scan.nextInt());
		int n = scan.nextInt();
		while (n-- > 0) {
			Target_Point pt = new Target_Point(scan.nextInt(), scan.nextInt());
			int count = 0;
			for (Shape sh : s)
				if (sh.isInside(pt))
					count++;
			System.out.println(count);
		}
		scan.close();
	}
}

interface Shape {
	boolean isInside(Target_Point pt);
}

class Target_Point {
	int x, y;

	public Target_Point(int x, int y) {
		this.x = x;
		this.y = y;
	}

	public double getDistanceto(Target_Point pt) {
		int _x = x - pt.x;
		int _y = y - pt.y;
		return Math.sqrt(_x * _x + _y * _y);
	}
}

class Rectangle implements Shape {
	Target_Point pt1, pt2;

	public Rectangle(int x1, int y1, int x2, int y2) {
		pt1 = new Target_Point(x1, y1);
		pt2 = new Target_Point(x2, y2);
	}

	public boolean isInside(Target_Point pt) {
		return pt.x >= pt1.x && pt.y >= pt1.y && pt.x <= pt2.x && pt.y <= pt2.y;
	}
}

class Circle implements Shape {
	Target_Point centre;
	int radius;

	public Circle(int x, int y, int radius) {
		centre = new Target_Point(x, y);
		this.radius = radius;
	}

	public boolean isInside(Target_Point pt) {
		return centre.getDistanceto(pt) <= radius;
	}
}

File: ts/src/hittingtargets.ts
import readline = require('readline')

interface Point {
  x: number
  y: number
}

interface Shape {
  isInside(pt: Point): boolean
}

class Circle implements Shape {
  // eslint-disable-next-line no-useless-constructor
  constructor (public center: Point, public radius: number) {}

  isInside (pt: Point): boolean {
    return (
      Math.pow(pt.x - this.center.x, 2) + Math.pow(pt.y - this.center.y, 2) <=
      Math.pow(this.radius, 2)
    )
  }
}

class Rectangle implements Shape {
  // eslint-disable-next-line no-useless-constructor
  constructor (public topLeft: Point, public bottomRight: Point) {}

  isInside (pt: Point): boolean {
    return (
      pt.x >= this.topLeft.x &&
      pt.x <= this.bottomRight.x &&
      pt.y >= this.topLeft.y &&
      pt.y <= this.bottomRight.y
    )
  }
}

let lineCount = 0
let targetCount = 0
const targets: Shape[] = []

readline
  .createInterface({
    input: process.stdin,
    output: process.stdout
  })
  .on('line', (line: string) => {
    if (lineCount === 0) {
      targetCount = parseInt(line)
    } else if (lineCount <= targetCount) {
      const lineArray = line.split(' ')
      targets.push(
        lineArray[0] === 'circle'
          ? new Circle(
            {
              x: parseInt(lineArray[1] ?? '0'),
              y: parseInt(lineArray[2] ?? '0')
            },
            parseInt(lineArray[3] ?? '0')
          )
          : new Rectangle(
            {
              x: parseInt(lineArray[1] ?? '0'),
              y: parseInt(lineArray[2] ?? '0')
            },
            {
              x: parseInt(lineArray[3] ?? '0'),
              y: parseInt(lineArray[4] ?? '0')
            }
          )
      )
    } else {
      if (!Number(line)) {
        const [x, y] = line.split(' ').map(Number)
        const pt: Point = { x: x ?? 0, y: y ?? 0 }
        console.log(`${targets.filter(target => target.isInside(pt)).length}`)
      }
    }
    lineCount++
  })
  .on('close', () => {
    process.exit(0)
  })
